### PR TITLE
python: Handle early errors during cockpit.spawn

### PR DESCRIPTION
**Pull request by mvollmer**
- [x] #17758 
- [x] land #17763 and rebase

### CODE REVIEW

File: pkg/base1/test-spawn-proc.js [hunk 1]
```diff
@@ -91,6 +91,32 @@ QUnit.test("error message fail", function (assert) {
             });
 });
 
+QUnit.test("nonexisting executable", function (assert) {
+    const done = assert.async();
+    assert.expect(2);
+    cockpit.spawn(["/bin/nonexistent"])
```
Comment (allisonkarlitskaya) [inline pkg/base1/test-spawn-proc.js:97]: Please rewrite this using async/await syntax.

File: pkg/base1/test-spawn-proc.js [hunk 2]
```diff
@@ -91,6 +91,32 @@ QUnit.test("error message fail", function (assert) {
...
+                assert.equal(this.state(), "rejected", "failed");
+                done();
+            });
+});
+
+QUnit.test("permission denied", function (assert) {
+    const done = assert.async();
+    assert.expect(2);
+    cockpit.spawn(["/etc/hostname"])
+            .fail(function(error) {
```
Comment (allisonkarlitskaya) [inline pkg/base1/test-spawn-proc.js:111]: ditto.

File: src/cockpit/channels/stream.py [hunk 3]
```diff
@@ -94,13 +94,26 @@ async def stream_process(self, args, options):
...
+        try:
+            process = await asyncio.create_subprocess_exec(
+                *args,
+                cwd=cwd,
+                env=env,
+                stdin=asyncio.subprocess.PIPE,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=stderr)
+        except FileNotFoundError:
+            problem = "not-found"
```
Comment (allisonkarlitskaya) [inline src/cockpit/channels/stream.py:107]: I do not like it.  Let me propose something that might help.

File: pkg/base1/test-spawn-proc.js (hunk 1 above)
Comment (mvollmer) [reply pkg/base1/test-spawn-proc.js:97]: I think this is actually best expressed with `assert.rejects`. WDYT?

**Comment by mvollmer** [comment]
> test-spawn-proc unfortunately hangs on "stream large output". Let's try to fix that before merging this.

https://github.com/cockpit-project/cockpit/pull/17758

File: pkg/base1/test-spawn-proc.js (hunk 1 above)
Comment (allisonkarlitskaya) [reply pkg/base1/test-spawn-proc.js:97]: I agree.

File: src/cockpit/channels/stream.py [hunk 4]
```diff
@@ -94,13 +94,21 @@ async def stream_process(self, args, options):
...
+                cwd=cwd,
+                env=env,
+                stdin=asyncio.subprocess.PIPE,
+                stdout=asyncio.subprocess.PIPE,
+                stderr=stderr)
+        except FileNotFoundError as error:
+            raise ChannelError('not-found') from error
+        except PermissionError as error:
+            raise ChannelError('access-denied') from error
+        except Exception as error:
```
Comment (allisonkarlitskaya) [inline src/cockpit/channels/stream.py:109]: The only question I have: I wonder if this should rather be OSError.

**Review by allisonkarlitskaya — APPROVED** [review]
Can land like this, but see one suggestion/comment below.

File: src/cockpit/channels/stream.py (hunk 4 above)
Comment (mvollmer) [reply src/cockpit/channels/stream.py:109]: Right, I think "yes".

**Review by allisonkarlitskaya — APPROVED** [review]
:+1: thanks!

**Review by allisonkarlitskaya — APPROVED** [review]
[check, check, checkity check, check](https://www.youtube.com/watch?v=_lYdxQ6vDGQ&t=17s)

**Comment by mvollmer** [comment]
Let's ignore the rhel-8-8 flake.

**Pull request merged**